### PR TITLE
Use hosts from site configurations for domain restriction

### PR DESCRIPTION
**Pull request by jonakieling**
Hi there,

we currently have a project where we try to remove the need for sys_domain records. One last thing was the host validation of this extension using only sys_domain records.

- This PR adds a field to the configuration model for storing site configurations.

- The hosts of those site configurations are then included in the list of domains checked by the extensions AuthenticationService.

- The TYPO3 Core API SiteFinder is used to get the site configurations and resolve the currently used variant and its host.

I'm very much open for suggestions regarding the changes. :)

kind regards,
jona

### CODE REVIEW

**Comment by stat1x** [comment]
Hi,
Would it not be better to have the LDAP configuration in the "Sites" module directly?

**Comment by jonakieling** [comment]
Hey,

the solution provided by the PR is mainly motivated by its simplicity since it uses the already existing code for the domain restrictions.
But I like the idea of having the LDAP config that is somewhat attached to a domain in the site config. As of right now I'm not too familiar with this extensions code base so I find it difficult to guess the time needed to implement the site config variant of this PR. It would also be a separate PR.

Here are some pros and cons that come to mind:

pros
+ all the LDAP config and mappings are under version control when put in a sites config yaml
+ domain restrictions are obsolete since the LDAP config is explicit for each site
+ site variants and environment variables can be used for different connections and credentials

cons
- when using multiple sites with the same LDAP config, each site needs to have the config in its entirety
- probably more time consuming than just adding the site configs domains

The last one is a bit of a blocker right now. I'll keep an eye on this and maybe find some time in the future for some coding. Right now we are using the current release since we have the same issue with the current ext:solr which only has PR for adding the solr config to the site config for a later release.

regards,
jona

**Comment by stat1x** [comment]
Thanks for the analysis. I agree with your pros/cons. Maybe we need to wait for EXT:solr to find a solution for this and then see how they managed this.
Best,
Jay.

**Comment by jonakieling** [comment]
Hi @xperseguers ,

are there plans to support TYPO3 site configurations in a future version of the extension? Can

File: Classes/Library/Configuration.php [hunk 1]
```diff
@@ -15,6 +15,8 @@
 namespace Causal\IgLdapSsoAuth\Library;
 
 use TYPO3\CMS\Core\Database\ConnectionPool;
+ use TYPO3\CMS\Core\Exception\SiteNotFoundException;
```
Comment (xperseguers) [inline Classes/Library/Configuration.php]: leading blank space

File: Classes/Library/Configuration.php [hunk 2]
```diff
@@ -85,6 +87,19 @@ public static function initialize($mode, \Causal\IgLdapSsoAuth\Domain\Model\Conf
             static::$domains[] = $row['domainName'];
         }
 
+        $siteFinder = GeneralUtility::makeInstance(SiteFinder::class);
```
Comment (xperseguers) [inline Classes/Library/Configuration.php]: will crash on TYPO3 v8

File: Classes/Library/Configuration.php [hunk 3]
```diff
@@ -85,6 +87,19 @@ public static function initialize($mode, \Causal\IgLdapSsoAuth\Domain\Model\Conf
...
+        $siteIdentifiers = GeneralUtility::trimExplode(',', $configuration->getSites(), true);
+        foreach ($siteIdentifiers as $siteIdentifier) {
+            try {
+                $host = $siteFinder->getSiteByIdentifier($siteIdentifier)->getBase()->getHost();
+                static::$domains[] = $host;
+            } catch (SiteNotFoundException $e) {
+                // ignore invalid site identifiers
```
Comment (xperseguers) [inline Classes/Library/Configuration.php]: comments start with a capital letter

File: Classes/Utility/SiteConfigurationItemsProcFunc.php [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+ *
+ * For the full copyright and license information, please read the
+ * LICENSE.txt file that was distributed with this source code.
+ *
+ * The TYPO3 project - inspiring people to share!
+ */
+
+namespace Causal\IgLdapSsoAuth\Utility;
+
+
```
Comment (xperseguers) [inline Classes/Utility/SiteConfigurationItemsProcFunc.php]: useless blank line

File: Configuration/TCA/tx_igldapssoauth_config.php [hunk 5]
```diff
@@ -30,7 +30,7 @@
         '1' => [
             'showitem' => '
                     --div--;GENERAL,
-                        hidden, name, domains,
+                        hidden, name, domains, sites,
```
[the remark this answers is not in the document]
Comment (jonakieling) [reply Configuration/TCA/tx_igldapssoauth_config.php]: I'm not aware of any best practices for this. Would a variable with changing value depending on the TYPO3 Version be ok? Like this:

`hidden, name, domains' . $sitesField . ',`

File: Configuration/TCA/tx_igldapssoauth_config.php [hunk 6]
```diff
@@ -1,4 +1,6 @@
 <?php
+$sitesField = version_compare(TYPO3_version, '9.0', '>=') ? 'sites,' : '';
```
Comment (xperseguers) [inline Configuration/TCA/tx_igldapssoauth_config.php]: my code uses the comma afterwards...

File: Configuration/TCA/tx_igldapssoauth_config.php [hunk 7]
```diff
@@ -17,7 +19,7 @@
         'iconfile' => 'EXT:ig_ldap_sso_auth/Resources/Public/Icons/icon_tx_igldapssoauth_config.png',
     ],
     'interface' => [
-        'showRecordFieldList' => 'hidden, name, domains, sites,
+        'showRecordFieldList' => 'hidden, name, domains' . $sitesField . ',
```
Comment (xperseguers) [inline Configuration/TCA/tx_igldapssoauth_config.php]: .. but you expect it to be before here

File: Configuration/TCA/tx_igldapssoauth_config.php [hunk 8]
```diff
@@ -30,7 +32,7 @@
         '1' => [
             'showitem' => '
                     --div--;GENERAL,
-                        hidden, name, domains, sites,
+                        hidden, name, domains' . $sitesField . ',
```
Comment (xperseguers) [inline Configuration/TCA/tx_igldapssoauth_config.php]: and here

**Comment by xperseguers** [comment]
I have the feeling that we should merge domains and site configuration in the extension. domains are deprecated since TYPO3 v9. Some business logic uses domains internally for checking for matching of configuration record or not, we may possibly merge them (you may have a look at branch `feature/typo3-v10`).

File: Configuration/TCA/tx_igldapssoauth_config.php (hunk 6 above)
Comment (jonakieling) [reply Configuration/TCA/tx_igldapssoauth_config.php]: Oh boy. I have cleaned up the commits into just two and fixed this error.

**Comment by jonakieling** [comment]
I have cleaned up the commits and will look into the typo3-v10 branch with a clear mind and report back.

**Pull request closed without merging**